refaster-runner/src/main/java/tech/picnic/errorprone/refaster/runner/Refaster.java

```diff
@@ -146,16 +159,40 @@ private static Stream<Replacement> getReplacements(
     return description.fixes.stream().flatMap(fix -> fix.getReplacements(endPositions).stream());
   }
 
-  private static CodeTransformer createCompositeCodeTransformer(ErrorProneFlags flags) {
+  // XXX: Add a flag to disable the optimized `RefasterRuleSelector`. That would allow us to verify
+  // that we're not prematurely pruning rules.
+  private static RefasterRuleSelector createRefasterRuleSelector(ErrorProneFlags flags) {
     ImmutableListMultimap<String, CodeTransformer> allTransformers =
         CodeTransformers.getAllCodeTransformers();
-    return CompositeCodeTransformer.compose(
+    List<RefasterRule<?, ?>> refasterRules = new ArrayList<>();
+    collectRefasterRules(
         flags
             .get(INCLUDED_TEMPLATES_PATTERN_FLAG)
             .map(Pattern::compile)
             .<ImmutableCollection<CodeTransformer>>map(
                 nameFilter -> filterCodeTransformers(allTransformers, nameFilter))
-            .orElseGet(allTransformers::values));
+            .orElseGet(allTransformers::values),
+        refasterRules::add);
+    return RefasterRuleSelector.create(ImmutableList.copyOf(refasterRules));
+  }
+
+  private static void collectRefasterRules(
+      ImmutableCollection<CodeTransformer> transformers, Consumer<RefasterRule<?, ?>> sink) {
+    for (CodeTransformer t : transformers) {
+      collectRefasterRules(t, sink);
+    }
+  }
+
+  private static void collectRefasterRules(
+      CodeTransformer transformer, Consumer<RefasterRule<?, ?>> sink) {
+    if (transformer instanceof RefasterRule) {
+      sink.accept((RefasterRule<?, ?>) transformer);
+    } else if (transformer instanceof CompositeCodeTransformer) {
+      collectRefasterRules(((CompositeCodeTransformer) transformer).transformers(), sink);
+    } else {
+      throw new IllegalStateException(
+          String.format("Can't handle `CodeTransformer` of type '%s'", transformer.getClass()));
+    }
   }
```

Comment:
@ibabiankou I agree that with the PR as-is it was an improvement to move this `RefasterRule` extraction logic here, but the changes in #255 make the `CodeTransformer` -> `ImmutableSet<ImmutableSet<String>>` transformation less straightforward, so I'll move it back. Of course, idea on how to better do this are welcome!